src/main/java/net/dv8tion/jda/core/requests/restaction/pagination/PaginationAction.java

```diff
@@ -0,0 +1,307 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.dv8tion.jda.core.requests.restaction.pagination;
+
+import net.dv8tion.jda.core.JDA;
+import net.dv8tion.jda.core.requests.Request;
+import net.dv8tion.jda.core.requests.Response;
+import net.dv8tion.jda.core.requests.RestAction;
+import org.apache.http.util.Args;
+
+import java.util.*;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * {@link net.dv8tion.jda.core.requests.RestAction RestAction} specification used
+ * to retrieve entities for paginated endpoints (before, after, limit).
+ *
+ * @param  <M>
+ *         The current implementation used as chaining return value
+ * @param  <T>
+ *         The type of entity to paginate
+ *
+ * @since  3.1
+ * @author Florian Spieß
+ */
+public abstract class PaginationAction<T, M extends PaginationAction<T, M>> extends RestAction<List<T>> implements Iterable<T>
+{
+
+    protected final List<T> cached = new CopyOnWriteArrayList<>();
+    protected final int maxLimit;
+    protected final int minLimit;
+    protected final AtomicInteger limit;
+
+    /**
+     * Creates a new PaginationAction instance
+     *
+     * @param api
+     *        The current JDA instance
+     * @param maxLimit
+     *        The inclusive maximum limit that can be used in {@link #limit(int)}
+     * @param minLimit
+     *        The inclusive minimum limit that can be used in {@link #limit(int)}
+     * @param initialLimit
+     *        The initial limit to use on the pagination endpoint
+     */
+    public PaginationAction(JDA api, int minLimit, int maxLimit, int initialLimit)
+    {
+        super(api, null, null);
+        this.maxLimit = maxLimit;
+        this.minLimit = minLimit;
+        this.limit = new AtomicInteger(initialLimit);
+    }
+
+    /**
+     * Creates a new PaginationAction instance
+     * <br>This is used for PaginationActions that should not deal with
+     * {@link #limit(int)}
+     *
+     * @param api
+     *        The current JDA instance
+     */
+    public PaginationAction(JDA api)
+    {
+        super(api, null, null);
+        this.maxLimit = 0;
+        this.minLimit = 0;
+        this.limit = new AtomicInteger(0);
+    }
+
+    /**
+     * The current amount of cached entities for this PaginationAction
+     *
+     * @return int size of cached entities
+     */
+    public int recentSize()
+    {
+        return cached.size();
+    }
+
+    /**
+     * Whether the cache of this PaginationAction is empty.
+     * <br>Logically equivalent to {@code recentSize() == 0}.
+     *
+     * @return True, if no entities have been retrieve yet.
+     */
+    public boolean isEmpty()
```

Comment:
What is the point of this method and `recentSize`? Can't you just `getRecent` and check the values on that?